### ISSUE TITLE
Project Miller indices (directions or plane normals) into a fundamental sector defined by a point group symmetry

### ISSUE THREAD

**Issue by hakonanes** [pyxem/orix#169]
After #164 is in, I plan to extend `orix.vector.SphericalRegion` with functionality to project Miller indices into a fundamental sector, with the properties `vertices`, `center` (just the mean) and `edges` (so we can delineate the sector in a stereographic plot).

As `Symmetry.fundamental_sector()` works now, I get an empty set of vectors for m-3m:

```python
>>> from orix.quaternion.symmetry import Oh
>>> Oh.fundamental_sector()
Vector3d (0,)
[]
```

Instead, I would expect

```python
>>> fs = Oh.fundamental_sector()
>>> fs
SphericalRegion (3,)
[[ 1 -1  0]
 [-1  0  1]
 [ 0  1  0]]
>>> fs.vertices
Vector3d (3,)
[[0.     0.     1.    ]
 [0.5774 0.5774 0.5774]
 [0.7071 0.     0.7071]]
>>> fs.vertices.scatter()
```

which produces this figure

![fs_m3m](https://user-images.githubusercontent.com/12139781/114300486-d0a7fa80-9ac0-11eb-8f86-33641baa040f.png)

I have used `SphericalRegion` just a little, and would like another opinion on:
1. Whether this seems like a natural thing for `Symmetry.fundamental_sector()` to return instead of a set of `Vector3d`?
2. The vertices, edges and center are appropriate for this class?

**Comment by pc494** [comment]
`fundamental_sector` seems to exist to solve a single, specific problem, namely finding the sector formed by the combination of two symmetry elements. Given it's lack of documentation and buried nature in the code I'm not sure it's the best place to start. I basically believe it should be private method.

Instead, `OrientationRegion`:
https://github.com/pyxem/orix/blob/fd8d28e3dd145e9defb0a9102b30e3cdcd4481cc/orix/quaternion/orientation_region.py#L112

Might be an alternative. It is this which we use to sample subregions of SO3.

(what follows is broadly applicable to the design of IPZ mapping more generally, so the points brought up in #166 and #122 will remain relevant)

However, my general feeling is that using SO3 (3D rotations) to create grids on S2 (the surface of a sphere) is a bad idea. Although it is very likely that a vector transformed by a uniform sampling of SO3 within a fundamental zone will produce a stereographic triangle there is no guarantee it would be the triangle we want (there is freedom to choose both FZ's and stereographic triangles).

The route I prefer is:

Fallback case
------------------

Grid the entirety of S2 (not very hard) and then cut all the regions that do not lie inside the corners of a stereographic triangle (hard coded). I think this is basically implemented in diffsims:

https://github.com/pyxem/diffsims/blob/b239a7e83ee38751fdc58c9b9797ab58aeef4567/diffsims/generators/rotation_list_generators.py#L173

The key difference is the return object (`vector3d` in orix) but I actually don't think the coaxing would be that difficult

Special cases
------------------

I think @din14970 implemented cubic very nicely in diffsims, and recent comment make me think he's also done hexagonal?

Pseudocode
----------------

```
get_streographic_triangle(special_grid=None,crystal_system=None,resolution=1,meshing='UV_grid'):
     """
     special_grid : str 
            a special designed grid that respects certain aesthetic considerations 
            (eg. includes high symmetry poles and the arcs  between them)
     crystal_system : str 
             for none custom routes 
     resolution : float 

     meshing : 
            meshing on S2, only used if special_grid is None
     """ 
     if special_system is not None: 
            # stuff 
     elif crystal_system == 'cubic':
            # ask user if they have considered our special cubic grids 
     else: 
            # get mesh 
            # eliminate points outside 

     #coaxing code for vector3d / miller
     return
```

**Comment by din14970** [comment]
I would say I largely agree with @hakonanes : in my view `SphericalRegion` should just store the vertices, edges and mean. Perhaps it is possible to add a method like `get_grid` which would use the SO2 meshing from diffsims (could possibly be migrated to orix at some point even) to create a set of `Vector3d` to sample the region. Such a method could take the same parameters as the mesh generation algorithms and automatically crop it to the right size. Whatever one does however, it must remain possible to generate a list of Euler angles with phi_1 constrained to 0 for my template matching algorithm to work. There's multiple ways this can be improved from the current situation. Maybe vector3d could get a method that converts it to bunge angles with phi1=0, effectively this is just a conversion to spherical coordinates with r = 1. 

>  I think @din14970 implemented cubic very nicely in diffsims, and recent comment make me think he's also done hexagonal?

The grids I have made should be applicable to any of the crystal systems, they do not exclusively serve for cubic. These are just ways to mesh the sphere, which happens to be convenient using a cubic mesh construct around the sphere. However, as I think you are pointing out, for some crystal systems like hexagonal, these grids result in a "rough cut" along the diagonal edge of the FZ. 
![image](https://user-images.githubusercontent.com/25320842/119967318-831e1900-bfac-11eb-976c-de9d8f87288c.png)

For template matching, the perfect gridding of this zone is not so critical, but I can understand that for some applications a custom tailored grid to each crystal system should be considered. To fix this, I would have to look back at how the grids are prepared and see if I can do something custom for hexagonal and trigonal.

**Comment by hakonanes** [comment]
After having read your comments a few times... I agree with you both.

It wasn't clear from my issue (top comment). What I want to use the `FundamentalTriangle` class for is to, given a `Symmetry`, get the `center` (vector) from the mean of the triangle's `vertices` (vectors) and then find the highest dot product between the symmetrically equivalent vectors of any (direct/reciprocal lattice) vector and the triangle `center`. As a result, the vector should then be inside fundamental triangle of the point group. We can then e.g. colour the vector, which initially is a map direction (say Z) rotated by a map orientation, to colour a pixel in an inverse pole figure (IPF) map.

This operation is basically MTEX' `vector3d.project2fundamentalRegion` ([link](https://mtex-toolbox.github.io/FundamentalSector.html)). Note that MTEX calls this class `FundamentalSector`, that is why I thought that `Symmetry.fundamental_sector()` would give me the same thing as MTEX returns...

So, my question is: Do I make a new class called `FundamentalTriangle` or something similar, or expand upon `SphericalRegion`? Knowing this, I can make a PR so we can discuss further.

**Comment by din14970** [comment]
> So, my question is: Do I make a new class called FundamentalTriangle or something similar, or expand upon SphericalRegion? Knowing this, I can make a PR so we can discuss further.

I would be tempted to call this a sub-class, since it might be handy to store a reference inside to the symmetry? It maybe depends on whether SphericalRegion has different purposes and is ever not a FundamentalTriangle.

**Comment by hakonanes** [comment]
> I would be tempted to call this a sub-class, since it might be handy to store a reference inside to the symmetry? It maybe depends on whether SphericalRegion has different purposes and is ever not a FundamentalTriangle.

That's a good option. It must have a symmetry attached, yes. Perhaps triangle isn't the best name, thinking about e.g. C1 which isn't a triangle...

**Comment by pc494** [comment]
It seems MTEX distinguishes between `FundamentalRegion` (SO3) and `FundamnetalSector` (pole figures) - if we could follow that notation consistently I think we might have solved our problem.

**Comment by hakonanes** [comment]
Agree, sounds good @pc494!

**Comment by hakonanes** [comment]
I would like this `FundamentalSector` to be available as a property `Symmetry`. Can we rename the `Symmetry.fundamental_sector()` to `Symmetry.fundamental_region()`? I believe this will be in line with naming in MTEX, as you mention @pc494.

**Comment by hakonanes** [comment]
@pc494, could you give your view on my suggestion in the previous comment?

**Comment by pc494** [comment]
It's marginal, but I think I would prefer `fundamental_zone()` as this is the choice in most of the literature.

Moving away from "sector" is absolutely fine - however it does mean that our next release will be reasonably significantly breaking. This works well for me, but if we have crucial new features that we would like to get in soonish (ie. before people have time to make the relevant changes for a breaking change) it may be worth considering delaying the merge of #213. My instinct is that this discussion and the code associated permeate into most of our new functionality, and so doing it all at once is probably best.

Does that help @hakonanes?

**Comment by hakonanes** [comment]
You mean change `Symmetry.fundamental_sector()` to `Symmetry.fundamental_zone()`? Sounds good to me.

I suggest for v0.7:
1. Add a deprecation warning to the `Symmetry.fundamental_sector()` *method* suggesting users call `Symmetry.fundamental_zone()` instead.
2. Move code from `fundamental_sector()` to `fundamental_zone()`, and let `fundamental_sector()` call `fundamental_zone()`, like we've done for `EBSD.match_patterns()` -> `EBSD.dictionary_indexing()` in [kikuchipy](https://github.com/pyxem/kikuchipy/blob/master/kikuchipy/signals/ebsd.py#L921).
3. Add the `Symmetry.fundamental_sector` *property* in #213.
4. Remove the `Symmetry.fundamental_sector()` *method* in v0.8.

**Comment by pc494** [comment]
Not sure this will work as you can't have a method and property with the same name (https://stackoverflow.com/questions/39210250/python-classes-method-has-same-name-as-property) - think it's probably simpler to do the whole thing in one fell swoop with good documentation so that when users inevitably run

```python
Symmetry.fundamental_sector? 
```

they get an explanation. Could potentially mess around with having `fundamental_sector` become a method that knows how it's being called and responds accordingly - but we're still a 0.x code and I think the time could be better spent elsewhere.

**Comment by hakonanes** [comment]
> Not sure this will work as you can't have a method and property with the same name

Did not know that, thanks.

> think it's probably simpler to do the whole thing in one fell swoop with good documentation

Agree. I'll make a suggestion for the change in #213 next week after vacation.

**Comment by hakonanes** [comment]
This will be possible in a follow-up PR to #232.
